File: codes/Garnetwzy/138.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;

    Node() {}

    Node(int _val, Node* _next, Node* _random) {
        val = _val;
        next = _next;
        random = _random;
    }
};
*/
class Solution {
public:
    Node* copyRandomList(Node* head) {
        if(!head)
            return head;
        Node *newHead = new Node(0, NULL, NULL);
        Node *current = newHead;
        Node* p = head;
        map<Node *, Node *> hash;
        hash[p] = current;
        while(p) {
            current->val = p->val;
            if(p->next) {
               if(hash.find(p->next) != hash.end()) {
                    current->next = hash[p->next];            
                }else {
                    hash[p->next] = new Node(0, NULL, NULL);   
                    current->next = hash[p->next];
                } 
            }
            
            if(p->random) {
                if(hash.find(p->random) != hash.end()) {
                    current->random = hash[p->random];
                }else{
                    hash[p->random] = new Node(0, NULL, NULL);
                    current->random = hash[p->random];
                }
            }
            p = p->next;
            current = current->next;
        }
        return newHead;
    }
};

File: codes/HellDream/138.java
class Solution {
    public Node copyRandomList(Node head) {
        if(head==null) return null;
        Node dummy = new Node(0,null,null);
        Node curr = dummy;
        HashMap<Node, Node> map = new HashMap<>();
        while(head!=null){
            if(!map.containsKey(head)){
                map.put(head, new Node(head.val, null, null));
            }
            curr.next = map.get(head);
            if(head.random!=null){
                if(!map.containsKey(head.random)){
                    map.put(head.random, new Node(head.random.val, null, null));
                }
            }
            curr.next.random = map.get(head.random);
            head = head.next;
            curr = curr.next;
        }
        return dummy.next;
    }
}